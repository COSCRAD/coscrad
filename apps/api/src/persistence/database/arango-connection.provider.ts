import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Database } from 'arangojs';
import { Scheme } from '../../app/config/constants/Scheme';
import { isNullOrUndefined } from '../../domain/utilities/validation/is-null-or-undefined';
import { DTO } from '../../types/DTO';
import ArangoDatabaseConfiguration from './ArangoDatabaseConfiguration';
import { getAllArangoCollectionIDs } from './collection-references/ArangoCollectionId';
import { isArangoEdgeCollectionCollectionID } from './collection-references/ArangoEdgeCollectionId';
import DatabaseAlreadyInitializedError from './errors/DatabaseAlreadyInitializedError';
import DatabaseCannotBeDroppedError from './errors/DatabaseCannotBeDroppedError';
import DatabaseNotYetInitializedError from './errors/DatabaseNotYetInitializedError';
import canDatabaseBeDropped from './utilities/canDatabaseBeDropped';

// Alias for more clarity from the outside; TODO wrap `Database` with simpler API?
export type ArangoConnection = Database;

type Port = `${number}`;

const isPortRequired = (scheme: Scheme, port: `${number}`) => {
    if (scheme === 'http' && port === '80') return false;
    if (scheme === 'https' && port === '443') return false;
    return true;
};

const buildFullHostURL = (
    domain: string,
    scheme: Scheme = Scheme.https,
    port: Port = '443'
): string => `${scheme}://${domain}${isPortRequired(scheme, port) ? `:${port}` : ''}`;

type AdminDatabaseConnectionOptions = {
    shouldConnectToSystemDatabase: boolean;
};

@Injectable()
export class ArangoConnectionProvider {
    #connection: ArangoConnection;

    #databaseConfiguration: ArangoDatabaseConfiguration;

    isInitialized = false;

    public get databaseConfiguration() {
        return this.#databaseConfiguration.getConfig();
    }

    constructor(private configService: ConfigService) {
        this.setDatabaseConfiguration({
            dbName: this.configService.get<string>('ARANGO_DB_NAME'),
            dbRootPass: this.configService.get<string>('ARANGO_DB_ROOT_PASSWORD'),
            dbUser: this.configService.get<string>('ARANGO_DB_USER'),
            dbPass: this.configService.get<string>('ARANGO_DB_USER_PASSWORD'),
            dbHostUrl: buildFullHostURL(
                this.configService.get<string>('ARANGO_DB_HOST_DOMAIN', 'localhost'),
                this.configService.get<Scheme>('ARANGO_DB_HOST_SCHEME', Scheme.http),
                this.configService.get<number>('ARANGO_DB_HOST_PORT', 80).toString() as Port
            ),
        });

        console.log(`dbName: ${this.databaseConfiguration.dbName}, dbRootPass: ${this.databaseConfiguration.dbRootPass}, dbUser: ${this.databaseConfiguration.dbUser}, dbPass: ${this.databaseConfiguration.dbPass}, dbHostUrl: ${this.databaseConfiguration.dbHostUrl}`);
        

        const systemDB = this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        });

        // TODO why is `useDatabase` deprecated? can we use myDB.database("db_name")?
        const dbInstance = systemDB.database(this.databaseConfiguration.dbName);
        dbInstance.useBasicAuth(
            this.databaseConfiguration.dbUser,
            this.databaseConfiguration.dbPass
        );

        this.#connection = dbInstance;
    }

    /**
     * This async setup helper will
     * - Create the database (the one referenced in your .env via ConfigService)
     * only if this database does not exist
     * - Add any missing collections (but leave existing collections untouched)
     * and can only be run once.
     */
    async initialize(): Promise<void> {
        if (this.isInitialized) throw new DatabaseAlreadyInitializedError();

        await this.#createDatabaseIfNotExists();

        await this.#createAllMissingCollections();

        this.isInitialized = true;
    }

    getDatabaseName(): string {
        return this.databaseConfiguration.dbName;
    }

    getConnection(): ArangoConnection {
        if (!this.isInitialized)
            throw new DatabaseNotYetInitializedError('get an Arango connection');

        return this.#connection;
    }

    public setDatabaseConfiguration(config: DTO<ArangoDatabaseConfiguration>) {
        this.#databaseConfiguration = new ArangoDatabaseConfiguration({
            ...config,
        });
    }

    #getAdminDBConnection({
        shouldConnectToSystemDatabase,
    }: AdminDatabaseConnectionOptions): Database {
        const adminInstance = new Database({
            url: this.databaseConfiguration.dbHostUrl,
        });

        adminInstance.useBasicAuth('root', this.databaseConfiguration.dbRootPass);

        if (shouldConnectToSystemDatabase) {
            return adminInstance;
        }

        // TODO why is `useDatabase` deprecated? can we use myDB.database("db_name")?
        const dbInstance = adminInstance.database(this.databaseConfiguration.dbName);

        return dbInstance;
    }

    async #doesCollectionExist(collectionName: string): Promise<boolean> {
        const allCollections = await this.#connection
            .listCollections()
            .then((allCollectionMetadata) => allCollectionMetadata.map(({ name }) => name));

        const doesCollectionExist = allCollections.includes(collectionName);

        return doesCollectionExist;
    }

    async #createDatabaseIfNotExists(): Promise<void> {
        const databaseName = this.databaseConfiguration.dbName;

        if (await this.#doesDatabaseExist(databaseName)) return;

        const { dbUser } = this.#databaseConfiguration;

        await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        }).createDatabase(databaseName, {
            users: [{ username: dbUser }, { username: 'root' }],
        });
    }

    async #createAllMissingCollections(): Promise<void> {
        await Promise.all(
            getAllArangoCollectionIDs().map((collectionName) =>
                this.#createCollectionIfNotExists(collectionName)
            )
        );
    }

    async #createCollectionIfNotExists(collectionName: string): Promise<void> {
        const doesCollectionExist = await this.#doesCollectionExist(collectionName);

        if (doesCollectionExist) return;

        /**
         * TODO [https://www.pivotaltracker.com/story/show/182132515]
         * cleanup the references
         */
        if (isArangoEdgeCollectionCollectionID(collectionName)) {
            await this.#connection.createEdgeCollection(collectionName);
        } else {
            await this.#connection.createCollection(collectionName);
        }
    }

    #doesDatabaseExist = async (databaseName: string) => {
        if (isNullOrUndefined(databaseName)) return false;

        const doesDatabaseExist = await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        })
            .listDatabases()
            .then((allDatabaseNames) => allDatabaseNames.includes(databaseName));

        return doesDatabaseExist;
    };

    /**
     * **CAUTION** This method is only to be used in tests.
     */
    async dropDatabaseIfExists(): Promise<void> {
        const dbName = this.databaseConfiguration.dbName;

        const doesDatabaseExist = await this.#doesDatabaseExist(dbName);

        if (!doesDatabaseExist) return;

        // for safety, you can only drop a db whose name matches the test db name pattern
        if (!canDatabaseBeDropped(dbName)) throw new DatabaseCannotBeDroppedError(dbName);

        await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        }).dropDatabase(dbName);
    }
}
