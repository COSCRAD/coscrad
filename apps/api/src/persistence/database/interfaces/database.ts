import { ResourceId } from 'apps/api/src/domain/types/ResourceId';
import { Maybe } from 'apps/api/src/lib/types/maybe';

export interface IDatabase {
    fetchById: <TEntityDTO>(id: ResourceId, collectionName: string) => Promise<Maybe<TEntityDTO>>;

    fetchMany: <TEntityDTO>(collectionName: string) => Promise<Maybe<TEntityDTO[]>>;

    create: <TEntityDTO>(dto: TEntityDTO, collectionName: string) => Promise<void>;

    createMany: <TEntityDTO>(dto: TEntityDTO[], collectionName: string) => Promise<void>;

    getCount: (collectionName: string) => Promise<number>;

    update: <TEntityDTO>(id: ResourceId, dto: TEntityDTO, collectionName: string) => Promise<void>;
}
