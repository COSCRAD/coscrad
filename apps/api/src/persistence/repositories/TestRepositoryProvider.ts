import { Entity } from '../../domain/models/entity';
import { EntityType, InMemorySnapshot } from '../../domain/types/entityType';
import { DatabaseProvider } from '../database/database.provider';
import { getArangoCollectionIDFromEntityType } from '../database/get-arango-collection-ids';
import { RepositoryProvider } from './repository.provider';

export default class TestRepositoryProvider extends RepositoryProvider {
  constructor(databaseProvider: DatabaseProvider) {
    super(databaseProvider);
  }

  // TODO We should correlate entity type with TEntity here
  public async addEntitiesOfSingleType<TEntity extends Entity>(
    entityType: EntityType,
    entities: TEntity[]
  ): Promise<void> {
    await this.forEntity<TEntity>(entityType).createMany(entities);
  }

  // TODO fix types
  public async addEntitiesOfManyTypes(
    snapshot: InMemorySnapshot
  ): Promise<void> {
    const writePromises = Object.entries(snapshot).map(
      ([entityType, entityInstances]) =>
        this.addEntitiesOfSingleType(
          entityType as EntityType,
          entityInstances as any
        )
    );
  }

  public async deleteAllEntitiesOfGivenType(
    entityType: EntityType
  ): Promise<void> {
    await (
      await this.databaseProvider.getDBInstance()
    ).deleteAll(getArangoCollectionIDFromEntityType(entityType));
  }
}
