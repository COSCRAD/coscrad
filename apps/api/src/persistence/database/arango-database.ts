import { Database } from 'arangojs';
import { aql, AqlQuery } from 'arangojs/aql';
import { isArangoDatabase } from 'arangojs/database';
import { Environment } from '../../app/config/constants/Environment';
import { QueryOperator } from '../../domain/repositories/interfaces/QueryOperator';
import { ISpecification } from '../../domain/repositories/interfaces/specification.interface';
import { isAggregateId } from '../../domain/types/AggregateId';
import { HasAggregateId } from '../../domain/types/HasAggregateId';
import { InternalError } from '../../lib/errors/InternalError';
import { Maybe } from '../../lib/types/maybe';
import { isNotFound, NotFound } from '../../lib/types/not-found';
import { isArangoCollectionId } from './collection-references/ArangoCollectionId';
import buildArangoDocumentHandle from './utilities/buildArangoDocumentHandle';
import { DatabaseDTO } from './utilities/mapEntityDTOToDatabaseDTO';

type ArangoDTO<T> = T & {
    _key: string;
    _id: string;
};

const aqlQueryOperators = {
    [QueryOperator.equals]: '==',
} as const;

const interpretQueryOperatorForAQL = (operator: QueryOperator): string => {
    const lookupResult = aqlQueryOperators[operator];

    if (!lookupResult) throw new InternalError(`Failed to parse operator: ${operator} for AQL.`);

    return lookupResult;
};

export class ArangoDatabase {
    #db: Database;

    constructor(database: Database) {
        if (!isArangoDatabase(database))
            throw new Error('Cannot create an Arango Database from an invalid database connection');

        this.#db = database;
    }

    fetchById = async <TDatabaseDTO extends DatabaseDTO<HasAggregateId>>(
        id: string,
        collectionName: string
    ): Promise<Maybe<TDatabaseDTO>> => {
        if (!isAggregateId(id)) {
            throw new InternalError(`Arango cannot fetchById with invalid id: ${id}`);
        }

        if (!isArangoCollectionId(collectionName)) {
            throw new InternalError(
                `Arango cannot fetchById from an invalid collection: ${collectionName}`
            );
        }

        const allEntities = await this.fetchMany<TDatabaseDTO>(collectionName);

        if (allEntities.length === 0) return NotFound;

        const searchId = buildArangoDocumentHandle(collectionName, id);

        const doIdsMatch = (searchId) => (dbDTO) => {
            const result = dbDTO._id === searchId;

            return result;
        };

        const searchResult = allEntities.find(doIdsMatch(searchId));

        return searchResult || NotFound;
    };

    /**
     *
     * @param collectionName name of the collection
     * @returns array of `DTOs`, empty array if none found
     */
    fetchMany = async <TEntityDTO>(
        collectionName: string,
        specification?: ISpecification<TEntityDTO>
    ): Promise<TEntityDTO[]> => {
        const { query: filterQuery, bindVars: filterBindVars } = specification
            ? this.#convertSpecificationToAQLFilter(specification, 't')
            : aql``;

        const query = specification
            ? `
      FOR t IN @@collectionName 
        ${filterQuery}
        return t
      `
            : `FOR t IN @@collectionName 
      return t
    `;

        const bindVars = {
            ...filterBindVars,
            '@collectionName': collectionName,
        };

        const aqlQuery: AqlQuery = {
            query,
            bindVars,
        };

        const cursor = await this.#db.query(aqlQuery);

        if (cursor.count === 0) return [];

        return cursor.all();
    };

    getCount = async (collectionName: string): Promise<number> => {
        if (!isArangoCollectionId(collectionName)) {
            throw new InternalError(
                `Arango cannot count for collection with invalid collection name: ${collectionName}`
            );
        }

        const results = await this.fetchMany(collectionName);

        return isNotFound(results) ? 0 : results.length;
    };

    create = async <TEntityDTO>(dto: TEntityDTO, collectionName: string): Promise<void> => {
        /**
         * Although the caller should ensure this, it's nice to double check here
         * as a means of making sure our query isn't subject to injection.
         */
        if (!isArangoCollectionId(collectionName)) {
            throw new Error(`Cannot insert into invalid collection: ${collectionName}`);
        }

        const collectionExists = await this.#doesCollectionExist(collectionName);

        if (!collectionExists) throw new Error(`Collection ${collectionName} not found!`);

        const query = `
    INSERT @dto
        INTO @@collectionName
    `;

        const bindVars = {
            dto,
            '@collectionName': collectionName,
        };

        await this.#db.query({
            query,
            bindVars,
        });
    };

    createMany = async <TEntityDTO>(dtos: TEntityDTO[], collectionName: string): Promise<void> => {
        const collectionExists = await this.#doesCollectionExist(collectionName);

        if (!collectionExists) throw new Error(`Collection ${collectionName} not found!`);

        const query = `
    FOR dto IN @dtos
        INSERT dto
            INTO @@collectionName
    `;

        const bindVars = {
            dtos,
            '@collectionName': collectionName,
        };

        /**
         * Apparently, this is a leaky abstraction from Arango's implementation
         * of RocksDB.  Note that this only currently seems to affect `createMany`,
         * which is only used in test setup. We may need to use this option for
         * other queries if we hit this problem elsewhere. The tell-tale sign is
         * a `write-write conflict` error from Arango.
         *
         * For more information see [here](https://github.com/arangodb/arangodb/issues/9702).
         */
        const MAX_NUMBER_OF_RETRIES = 10;

        await this.#db.query(
            {
                query,
                bindVars,
            },
            {
                retryOnConflict: MAX_NUMBER_OF_RETRIES,
            }
        );
    };

    update = async <TUpdateEntityDTO>(
        id: string,
        updatedDto: TUpdateEntityDTO,
        collectionName: string
    ): Promise<void> => {
        const documentToUpdate = await this.fetchById(id, collectionName);

        if (isNotFound(documentToUpdate))
            throw new Error(
                `Cannot update document: ${id} in collection: ${collectionName}, as no document with that id was found`
            );

        // TODO remove cast
        const key = this.#getKeyOfDocument(
            documentToUpdate as unknown as ArangoDTO<TUpdateEntityDTO>
        );

        if (isNotFound(key))
            throw new Error(`No property '_key' was found on document: ${documentToUpdate}`);

        const query = `
            UPDATE @updatedDto IN @@collectionName OPTIONS { keepNull: false }
        `;

        const bindVars = {
            updatedDto: {
                /**
                 * An update query in arango must include enough information
                 * to uniquely identify the document to update.
                 *
                 * Further note that we are merging the `updatedDTO` with the
                 * existing document in case there is information that is persisted
                 * but not populated on the model (e.g. event history, metadata).
                 */
                _key: key,
                ...updatedDto,
            },
            '@collectionName': collectionName,
        };

        await this.#db
            .query({
                query,
                bindVars,
            })
            .catch((err) => {
                throw new InternalError(
                    `Failed to update dto: ${updatedDto} in ${collectionName}. \n Arango errors: ${err}`
                );
            });
    };

    updateMany = async <TUpdatedDoc extends { _key: string }>(
        docUpdates: TUpdatedDoc[],
        collectionName: string
    ): Promise<void> => {
        const query = `
            FOR docUpdate in @docUpdates
            UPDATE docUpdate IN @@collectionName OPTIONS { keepNull: false }
        `;

        const bindVars = {
            docUpdates,
            '@collectionName': collectionName,
        };

        await this.#db
            .query({
                query,
                bindVars,
            })
            .catch((err) => {
                throw new InternalError(
                    `Failed to update dto: ${JSON.stringify(
                        docUpdates
                    )} in ${collectionName}. \n Arango errors: ${err}`
                );
            });
    };

    // TODO Add Soft Delete

    // TODO we should only expose a hard delete for test setup
    delete = async (id: string, collectionName: string): Promise<void> => {
        const documentToRemove = await this.fetchById(id, collectionName);

        if (isNotFound(documentToRemove))
            throw new InternalError(
                `You cannot remove document ${id} in collection ${collectionName} as it does not exist`
            );

        throw new InternalError('ArangoDatabase.delete Not Implemented');
    };

    // TODO We only want this power within test utilities!
    deleteAll = async (collectionName: string): Promise<void> => {
        if (process.env.NODE_ENV !== Environment.test) {
            throw new InternalError(`You can only delete all in a test environment`);
        }

        if (!isArangoCollectionId(collectionName)) {
            throw new InternalError(`Cannot delete all in invalid collection: ${collectionName}`);
        }

        const query = `
    FOR doc in @@collectionName
    REMOVE doc in @@collectionName
    `;

        const bindVars = {
            '@collectionName': collectionName,
        };

        this.#db.query({ query, bindVars });
    };

    #getKeyOfDocument = <TEntityDTO>(document: ArangoDTO<TEntityDTO>): Maybe<string> =>
        typeof document._key === 'string' ? document._key : NotFound;

    #doesCollectionExist = async (collectionName: string): Promise<boolean> =>
        this.#db
            .collections()
            .then((collections) =>
                collections.some((collection) => collection.name === collectionName)
            );

    #convertSpecificationToAQLFilter<TModel>(
        { criterion: { field, operator, value } }: ISpecification<TModel>,
        docNamePlaceholder: string
    ) {
        return {
            query: `FILTER ${docNamePlaceholder}.${field} ${interpretQueryOperatorForAQL(
                operator
            )} @valueToCompare`,
            bindVars: {
                // This prevents AQL injection attacks
                valueToCompare: value,
            },
        };
    }
}
