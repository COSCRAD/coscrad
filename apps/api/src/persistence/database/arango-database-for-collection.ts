import { isArangoDatabase } from 'arangojs/database';
import { ISpecification } from '../../domain/repositories/interfaces/ISpecification';
import { AggregateId } from '../../domain/types/AggregateId';
import { HasAggregateId } from '../../domain/types/HasAggregateId';
import { Maybe } from '../../lib/types/maybe';
import { DeepPartial } from '../../types/DeepPartial';
import { ArangoDatabase } from './arango-database';
import { ArangoCollectionId } from './collection-references/ArangoCollectionId';
import { DatabaseDocument } from './utilities/mapEntityDTOToDatabaseDTO';

/**
 * Note that at this level we are working with a `DatabaseDocument` (has _key
 * and _id), not an `EntityDTO`. The mapping is taken care of in the
 * repositories layer.
 */
export class ArangoDatabaseForCollection<TEntity extends HasAggregateId> {
    #collectionID: ArangoCollectionId;

    #arangoDatabase: ArangoDatabase;

    constructor(arangoDatabase: ArangoDatabase, collectionName: ArangoCollectionId) {
        this.#collectionID = collectionName;

        this.#arangoDatabase = arangoDatabase;

        if (isArangoDatabase(this.#arangoDatabase))
            throw new Error(
                `Received invalid arango db instance: ${JSON.stringify(arangoDatabase)}`
            );
    }

    // Queries (return information)
    fetchById(id: AggregateId): Promise<Maybe<DatabaseDocument<TEntity>>> {
        return this.#arangoDatabase.fetchById<DatabaseDocument<TEntity>>(id, this.#collectionID);
    }

    fetchMany(specification?: ISpecification<TEntity>): Promise<DatabaseDocument<TEntity>[]> {
        return this.#arangoDatabase.fetchMany<DatabaseDocument<TEntity>>(
            this.#collectionID,
            // TODO remove cast, handle mapping layer
            specification as unknown as ISpecification<DatabaseDocument<TEntity>>
        );
    }

    getCount(): Promise<number> {
        return this.#arangoDatabase.getCount(this.#collectionID);
    }

    // Commands (mutate state)
    create(DatabaseDocument: DatabaseDocument<TEntity>) {
        // Handle the difference in _id \ _key between model and database
        return this.#arangoDatabase.create(DatabaseDocument, this.#collectionID);
    }

    createMany(DatabaseDocuments: DatabaseDocument<TEntity>[]) {
        return this.#arangoDatabase.createMany(DatabaseDocuments, this.#collectionID);
    }

    update(id: AggregateId, updateDTO: DeepPartial<DatabaseDocument<TEntity>>) {
        return this.#arangoDatabase.update(id, updateDTO, this.#collectionID);
    }
}
