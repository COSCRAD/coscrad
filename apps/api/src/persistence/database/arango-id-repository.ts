import { AggregateId } from '../../domain/types/AggregateId';
import { InternalError } from '../../lib/errors/InternalError';
import { IIdRepository } from '../../lib/id-generation/interfaces/id-repository.interface';
import { UuidDocument } from '../../lib/id-generation/types/UuidDocument';
import { Maybe } from '../../lib/types/maybe';
import { isNotFound } from '../../lib/types/not-found';
import { ArangoDatabase } from './arango-database';
import { ArangoCollectionId } from './collection-references/ArangoCollectionId';
import mapDatabaseDTOToEntityDTO from './utilities/mapDatabaseDTOToEntityDTO';
import mapEntityDTOToDatabaseDTO, { DatabaseDocument } from './utilities/mapEntityDTOToDatabaseDTO';

type IdDocument = DatabaseDocument<{ id: AggregateId; isAvailable: boolean }>;

export class ArangoIdRepository implements IIdRepository<AggregateId> {
    constructor(private readonly arangoDatabase: ArangoDatabase) {}

    async fetchById(id: AggregateId): Promise<Maybe<UuidDocument<string>>> {
        const result = await this.arangoDatabase.fetchById<
            DatabaseDocument<UuidDocument<AggregateId>>
        >(id, ArangoCollectionId.uuid);

        return isNotFound(result)
            ? result
            : mapDatabaseDTOToEntityDTO<UuidDocument<AggregateId>>(result);
    }

    async create(id: AggregateId): Promise<void> {
        await this.arangoDatabase.create(
            mapEntityDTOToDatabaseDTO({
                id,
                isAvailable: true,
            }),
            ArangoCollectionId.uuid
        );
    }

    async reserve(id: AggregateId): Promise<void> {
        const result = await this.fetchById(id);

        if (isNotFound(result)) {
            throw new InternalError(
                `Cannot reserve id: ${id}, as it has not been registered with our system`
            );
        }

        const { isAvailable } = result;

        if (!isAvailable) {
            throw new InternalError(`Cannot reserve id: ${id} as it is already in use`);
        }

        await this.arangoDatabase.update<Partial<IdDocument>>(
            id,
            {
                isAvailable: false,
            },
            ArangoCollectionId.uuid
        );
    }
}
