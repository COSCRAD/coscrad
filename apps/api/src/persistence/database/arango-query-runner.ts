import { HasId } from '@coscrad/api-interfaces';
import { Injectable } from '@nestjs/common';
import { isDeepStrictEqual } from 'util';
import { ICoscradQueryRunner } from '../migrations/coscrad-query-runner.interface';
import { ArangoDatabase } from './arango-database';
import { ArangoCollectionId } from './collection-references/ArangoCollectionId';
import { ArangoDatabaseProvider } from './database.provider';
import { ArangoDatabaseDocument } from './utilities/mapEntityDTOToDatabaseDTO';

@Injectable()
export class ArangoQueryRunner implements ICoscradQueryRunner {
    private readonly arangoDatabase: ArangoDatabase;

    constructor(arangoDatabaseProvider: ArangoDatabaseProvider) {
        this.arangoDatabase = arangoDatabaseProvider.getDBInstance();
    }

    async update<TOldDocument extends ArangoDatabaseDocument<HasId>, UNewDocument>(
        collectionName: ArangoCollectionId,
        calculateUpdate: (oldDocument: TOldDocument) => UNewDocument
    ): Promise<void> {
        console.log(`running update queries with: ${this.arangoDatabase.getDatabaseName()}`);

        const existingDocs = await this.arangoDatabase.fetchMany<TOldDocument>(collectionName);

        const updates = existingDocs
            .map((oldDocument) => ({
                ...calculateUpdate(oldDocument),
                // this ensures that the update can be matched to an existing doc
                _key: oldDocument._key,
            }))
            // avoid unnecessary writes
            .filter((update) => !isDeepStrictEqual(update, {}));

        await this.arangoDatabase.updateMany(updates, collectionName);
    }
}
