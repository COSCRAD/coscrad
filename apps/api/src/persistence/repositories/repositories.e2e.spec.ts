import { INestApplication } from '@nestjs/common';
import createTestModule from '../../app/controllers/__tests__/createTestModule';
import getInstanceFactoryForEntity from '../../domain/factories/getInstanceFactoryForEntity';
import { Entity } from '../../domain/models/entity';
import { entityTypes } from '../../domain/types/entityTypes';
import { InternalError, isInternalError } from '../../lib/errors/InternalError';
import { NotFound } from '../../lib/types/not-found';
import buildTestData from '../../test-data/buildTestData';
import { ArangoConnectionProvider } from '../database/arango-connection.provider';
import { DatabaseProvider } from '../database/database.provider';
import generateRandomTestDatabaseName from './__tests__/generateRandomTestDatabaseName';
import TestRepositoryProvider from './__tests__/TestRepositoryProvider';

describe('Repository provider > repositoryForEntity', () => {
    const testDatabaseName = generateRandomTestDatabaseName();

    const testData = buildTestData();

    let arangoConnectionProvider: ArangoConnectionProvider;

    let databaseProvider: DatabaseProvider;

    let testRepositoryProvider: TestRepositoryProvider;

    let app: INestApplication;

    beforeAll(async () => {
        jest.resetModules();

        const moduleRef = await createTestModule(testDatabaseName);

        arangoConnectionProvider =
            moduleRef.get<ArangoConnectionProvider>(ArangoConnectionProvider);

        databaseProvider = new DatabaseProvider(arangoConnectionProvider);

        testRepositoryProvider = new TestRepositoryProvider(databaseProvider);

        app = moduleRef.createNestApplication();

        await app.init();
    });

    afterAll(async () => {
        await testRepositoryProvider.testTeardown();

        await app.close();
    });

    Object.values(entityTypes).forEach((entityType) => {
        describe(`Repository for entity of type ${entityType}`, () => {
            beforeEach(async () => {
                await testRepositoryProvider.addEntitiesOfManyTypes(testData);
            });

            afterEach(async () => {
                await testRepositoryProvider.testTeardown();
            });
            describe('fetchMany', () => {
                describe('When no specification is provided', () => {
                    it('should return the expected result', async () => {
                        const result = await testRepositoryProvider
                            .forEntity(entityType)
                            .fetchMany();

                        /**
                         * TODO [https://www.pivotaltracker.com/story/show/181503421] setup a
                         * custom matcher in Jest for comparing instances (not DTOs)
                         * */
                        expect(JSON.stringify(result)).toEqual(
                            JSON.stringify(testData[entityType])
                        );
                    });
                });
            });

            describe('fetchById', () => {
                describe('when there is an entity with the given id', () => {
                    it('should return the expected result', async () => {
                        const entityToFind = testData[entityType][0];

                        const result = await testRepositoryProvider
                            .forEntity(entityType)
                            .fetchById(entityToFind.id);

                        // TODO custom matcher (Same as above)
                        expect(JSON.stringify(result)).toEqual(JSON.stringify(entityToFind));
                    });
                });

                describe('when there is no entity with the given id', () => {
                    it('should return not found', async () => {
                        const result = await testRepositoryProvider
                            .forEntity(entityType)
                            .fetchById('BOGUS-ENTITY-ID');

                        // TODO custom matcher (Same as above)
                        expect(result).toEqual(NotFound);
                    });
                });
            });

            //
            describe('getCount', () => {
                it('should return the expected count', async () => {
                    const expectedCount = testData[entityType].length;

                    const actualCount = await testRepositoryProvider
                        .forEntity(entityType)
                        .getCount();

                    expect(actualCount).toEqual(expectedCount);
                });
            });

            describe('create', () => {
                it('should successfully create the new entity', async () => {
                    const dtoForEntityToCreate = {
                        ...testData[entityType][0].toDTO(),
                        id: 'BRAND-NEW-ENTITY-ID',
                    };

                    const entityFactory = getInstanceFactoryForEntity(entityType);

                    const newEntityInstance = entityFactory(dtoForEntityToCreate);

                    /**
                     * This is just to satisfy the typeScript compiler. It shouldn't happen,
                     * as we already pass our test data through the validators as part of
                     * a separate test.
                     */
                    if (isInternalError(newEntityInstance)) throw newEntityInstance;

                    await testRepositoryProvider.forEntity(entityType).create(newEntityInstance);

                    const entityFetchedAfterCreation = await testRepositoryProvider
                        .forEntity(entityType)
                        .fetchById(newEntityInstance.id);

                    expect(entityFetchedAfterCreation).not.toBe(NotFound);

                    expect(JSON.stringify(entityFetchedAfterCreation)).toEqual(
                        JSON.stringify(newEntityInstance)
                    );
                });
            });

            describe('createMany', () => {
                it('should successfully create all new entities', async () => {
                    const entityFactory = getInstanceFactoryForEntity(entityType);

                    const newEntitiesToCreateOrErrors = testData[entityType]
                        .map((oldEntity, index) => ({
                            ...oldEntity.toDTO(),
                            id: `NEW-ENTITY-ID-${index + 1}`,
                        }))
                        .map((dto) => entityFactory(dto));

                    /**
                     * **note** This shouldn't happen as we already test that our test
                     * data satisfies invariant validation and we are only changing the
                     * id here (so to avoid collisions in the db).
                     */
                    if (newEntitiesToCreateOrErrors.some(isInternalError))
                        throw new InternalError(
                            [
                                `Encountered invalid test data for entity of type: ${entityType}.`,
                                `\n You may want to run validateTestData.spec.ts`,
                            ].join(' ')
                        );

                    /**
                     * This is solely to quiet typeCheck without casting, as we've
                     * already thrown if any validation errors were returned from the
                     * factory.
                     */
                    const newEntitiesToCreate = newEntitiesToCreateOrErrors.filter(
                        (entityOrError): entityOrError is Entity => !isInternalError(entityOrError)
                    );

                    await testRepositoryProvider
                        .forEntity(entityType)
                        .createMany(newEntitiesToCreate);

                    newEntitiesToCreate.forEach(async (entity) => {
                        const entityFetchedAfterCreation = await testRepositoryProvider
                            .forEntity(entityType)
                            .fetchById(entity.id);

                        expect(entityFetchedAfterCreation).not.toBe(NotFound);

                        expect(JSON.stringify(entityFetchedAfterCreation)).toEqual(
                            JSON.stringify(entity)
                        );
                    });
                });
            });
        });
    });
});
