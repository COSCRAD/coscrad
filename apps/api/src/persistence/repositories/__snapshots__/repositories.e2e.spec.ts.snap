// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /entities (fetch view models)- all entities published when querying for view models for entity of type tag when an id is not provided ?type=tag (fetching many view models) - all entities published should fetch multiple entities of type tag 1`] = `
Array [
  Object {
    "id": "0",
    "text": "plants",
  },
  Object {
    "id": "1",
    "text": "animals",
  },
  Object {
    "id": "2",
    "text": "placenames",
  },
  Object {
    "id": "3",
    "text": "songs",
  },
  Object {
    "id": "4",
    "text": "legends",
  },
]
`;

exports[`GET /entities (fetch view models)- all entities published when querying for view models for entity of type term when an id is not provided ?type=term (fetching many view models) - all entities published should fetch multiple entities of type term 1`] = `
Array [
  Object {
    "contributor": "",
    "id": "1",
    "isPublished": true,
    "term": "Chil-term-1",
    "termEnglish": "Engl-term-1",
  },
  Object {
    "contributor": "",
    "id": "2",
    "isPublished": true,
    "term": "Chil-term-2",
    "termEnglish": "Engl-term-2",
  },
  Object {
    "contributor": "",
    "id": "3",
    "isPublished": true,
    "term": "Chil-term-no-english",
  },
]
`;

exports[`GET /entities (fetch view models)- all entities published when querying for view models for entity of type vocabularyList when an id is not provided ?type=vocabularyList (fetching many view models) - all entities published should fetch multiple entities of type vocabularyList 1`] = `
Array [
  Object {
    "entries": Array [
      Object {
        "term": Object {
          "contributor": "",
          "id": "1",
          "isPublished": true,
          "term": "Chil-term-1",
          "termEnglish": "Engl-term-1",
        },
        "variableValues": Object {
          "person": "11",
        },
      },
      Object {
        "term": Object {
          "contributor": "",
          "id": "2",
          "isPublished": true,
          "term": "Chil-term-2",
          "termEnglish": "Engl-term-2",
        },
        "variableValues": Object {
          "person": "21",
        },
      },
    ],
    "id": "vocabulary-list-id-1",
    "isPublished": true,
    "name": "test VL 1 chil",
    "nameEnglish": "test VL 1 engl",
    "variables": Array [
      Object {
        "name": "person",
        "type": "dropbox",
        "validValues": Array [
          Object {
            "display": "I",
            "value": "11",
          },
          Object {
            "display": "We",
            "value": "21",
          },
        ],
      },
    ],
  },
  Object {
    "entries": Array [
      Object {
        "term": Object {
          "contributor": "",
          "id": "2",
          "isPublished": true,
          "term": "Chil-term-2",
          "termEnglish": "Engl-term-2",
        },
        "variableValues": Object {
          "person": "23",
        },
      },
    ],
    "id": "vocabulary-list-id-2",
    "isPublished": true,
    "name": "test VL 2 CHIL- no engl name",
    "variables": Array [],
  },
]
`;
